File: src/main/java/net/serebryansky/city/controller/CityController.java
package net.serebryansky.city.controller;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import net.serebryansky.city.model.City;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Mono;

import static net.serebryansky.common.util.LoggingUtil.logOnEach;
import static org.springframework.http.ResponseEntity.ok;

@RestController
@RequestMapping("cities")
@Slf4j
@RequiredArgsConstructor
public class CityController {
    @GetMapping("{id}")
    public ResponseEntity<Mono<City>> get(@PathVariable String id) {
        val city = new City();
        city.setId(id);
        city.setName("City #" + id);
        return ok(Mono.empty()
                .doOnEach(logOnEach(() -> log.info("Get city #{}", id)))
                .then(Mono.just(city)));
    }
}

File: src/main/kotlin/net/serebryansky/city/controller/CityController.kt
package net.serebryansky.city.controller

import kotlinx.coroutines.reactive.awaitFirst
import mu.KotlinLogging
import net.serebryansky.city.model.City
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Mono

@RestController
@RequestMapping("cities")
class CityController {

    private val log = KotlinLogging.logger {}

    @GetMapping("{id}")
    suspend fun get1(@PathVariable id: String): City {
        log.info { "Get city $id" }
        val city = Mono.just(City(
                id = id,
                name = "City #$id")
        )
        return city.awaitFirst()
    }
}
